src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java

```diff
@@ -1534,12 +1537,28 @@ static void parseKeyStoreType(ConfigurationXMLStreamReader reader, final Map<Str
 
     private static ExceptionSupplier<CredentialSource, ConfigXMLParseException> createCredentialStoreSupplier(final XMLLocation location, final String storeName, final String alias, final String clearText, final Map<String, ExceptionSupplier<CredentialStore, ConfigXMLParseException>> credentialStoresMap) {
         return () -> {
-            final ExceptionSupplier<CredentialStore, ConfigXMLParseException> supplier = credentialStoresMap.get(storeName);
-            if (supplier == null) {
-                throw xmlLog.xmlCredentialStoreNameNotDefined(location, storeName);
+            if (storeName != null) {
+                final ExceptionSupplier<CredentialStore, ConfigXMLParseException> supplier = credentialStoresMap.get(storeName);
+                if (supplier == null) {
+                    throw xmlLog.xmlCredentialStoreNameNotDefined(location, storeName);
+                }
+                final CredentialStore credentialStore = supplier.get();
+                return new CredentialStoreCredentialSource(credentialStore, alias);
+            } else if (clearText != null) {
+                return new CredentialSource() {
+                    @Override
+                    public SupportLevel getCredentialAcquireSupport(Class<? extends Credential> credentialType, String algorithmName, AlgorithmParameterSpec parameterSpec) throws IOException {
+                        return SupportLevel.SUPPORTED;
+                    }
+
+                    @Override
+                    public <C extends Credential> C getCredential(Class<C> credentialType, String algorithmName, AlgorithmParameterSpec parameterSpec) throws IOException {
+                        return credentialType.cast(new PasswordCredential(ClearPassword.createRaw(ClearPassword.ALGORITHM_CLEAR, clearText.toCharArray())));
```

Comment:
This can cause a ClassCastException if (for some reason) the user of this CredentialSource tries to query some other credential type.  A better option is to:

* Create the PasswordCredential outside the anonymous class to avoid allocation pressure, and
* Use ```return passwordCredential.castAs(credentialType, algorithmName, parameterSpec);``` as the body, which will return ```null``` for any unsupported requests as per spec.